src/mesh/p4est_mesh.jl

```diff
@@ -403,6 +609,66 @@ function map_node_coordinates!(node_coordinates::AbstractArray{<:Any, 4}, mappin
   return node_coordinates
 end
 
+# 3D version
+function calc_tree_node_coordinates!(node_coordinates::AbstractArray{RealT, 5},
+                                     nodes, mapping,
+                                     vertices, tree_to_vertex) where RealT
+  nodes_in = [-1.0, 1.0]
+  matrix = polynomial_interpolation_matrix(nodes_in, nodes)
+  data_in = Array{RealT, 4}(undef, 3, 2, 2, 2)
+
+  for tree in 1:size(tree_to_vertex, 2)
+    # Tree vertices are stored in Z-order, zero-based indexing
+    data_in[:, 1, 1, 1] .= vertices[:, tree_to_vertex[1, tree] + 1]
+    data_in[:, 2, 1, 1] .= vertices[:, tree_to_vertex[2, tree] + 1]
+    data_in[:, 1, 2, 1] .= vertices[:, tree_to_vertex[3, tree] + 1]
+    data_in[:, 2, 2, 1] .= vertices[:, tree_to_vertex[4, tree] + 1]
+    data_in[:, 1, 1, 2] .= vertices[:, tree_to_vertex[5, tree] + 1]
+    data_in[:, 2, 1, 2] .= vertices[:, tree_to_vertex[6, tree] + 1]
+    data_in[:, 1, 2, 2] .= vertices[:, tree_to_vertex[7, tree] + 1]
+    data_in[:, 2, 2, 2] .= vertices[:, tree_to_vertex[8, tree] + 1]
```

Comment:
This will allocate (RHS). Maybe you could use `@view`s? On the other hand, it's only used once in the constructor, so it's probably irrelevant...